omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/ModuleActionResource1_8.java

```diff
@@ -0,0 +1,245 @@
+import org.openmrs.module.webservices.rest.web.RequestContext;
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.annotation.Resource;
+import org.openmrs.module.webservices.rest.web.representation.Representation;
+import org.openmrs.module.webservices.rest.web.resource.api.Creatable;
+import org.openmrs.module.webservices.rest.web.resource.impl.BaseDelegatingResource;
+import org.openmrs.module.webservices.rest.web.resource.impl.DelegatingResourceDescription;
+import org.openmrs.module.webservices.rest.web.response.ResourceDoesNotSupportOperationException;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+
+import javax.servlet.ServletContext;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+@Resource(name = RestConstants.VERSION_1 + "/moduleaction", supportedClass = ModuleAction.class, supportedOpenmrsVersions = {
+        "1.8.*", "1.9.*", "1.10.*", "1.11.*", "1.12.*", "2.0.*", "2.1.*" })
+public class ModuleActionResource1_8 extends BaseDelegatingResource<ModuleAction> implements Creatable {
+	
+	private ModuleFactoryWrapper moduleFactoryWrapper = new ModuleFactoryWrapper();
+	
+	public void setModuleFactoryWrapper(ModuleFactoryWrapper moduleFactoryWrapper) {
+		this.moduleFactoryWrapper = moduleFactoryWrapper;
+	}
+	
+	/**
+	 * Overriding create directly, because ModuleFactory requires ServletContext to execute any
+	 * action
+	 */
+	@Override
+	public Object create(SimpleObject post, RequestContext context) throws ResponseException {
+		moduleFactoryWrapper.checkPrivilege();
+		ModuleAction action = parseAction(post, context);
+		Collection<Module> modules = action.isAllModules() ? moduleFactoryWrapper.getLoadedModules() : action.getModules();
+		
+		ServletContext servletContext = getServletContext(context);
+		
+		if (modules.isEmpty()) {
+			throw new IllegalArgumentException("Cannot execute action " + action.getType() + " on empty set of modules.");
+		} else {
+			switch (action.getType()) {
+				case START:
+					startModules(modules, servletContext);
+					break;
+				case STOP:
+					stopModules(modules, servletContext);
+					break;
+				case RESTART:
+					restartModules(modules, servletContext);
+					break;
+				case UNLOAD:
+					unloadModules(modules, servletContext);
+					break;
+			}
+		}
+		
+		return ConversionUtil.convertToRepresentation(action, Representation.DEFAULT);
+	}
+	
+	private ModuleAction parseAction(SimpleObject post, RequestContext context) {
+		
+		//parse action
+		String actionTypeName = post.get("action");
+		ModuleAction.Type actionType;
+		try {
+			if (actionTypeName == null) {
+				throw new IllegalArgumentException("Please specify action type");
+			}
+			actionType = ModuleAction.Type.valueOf(actionTypeName.toUpperCase().trim());
+		}
+		catch (Exception e) {
+			throw new IllegalArgumentException("Invalid modules action", e);
```

Comment:
Throw IllegalRequestException. No need to rethrow "Please specify action type", make the try block smaller.